src/analyses/mallocFresh.ml

```diff
@@ -6,34 +6,38 @@ module Spec =
 struct
   include Analyses.IdentitySpec
 
-  (* must fresh variables *)
-  module D = SetDomain.Reverse (SetDomain.ToppedSet (CilType.Varinfo) (struct let topname = "All variables" end)) (* need bot (top) for hoare widen *)
+  (* must fresh (or may not fresh) variables *)
+  module DF = SetDomain.ToppedSet (CilType.Varinfo) (struct let topname = "All variables" end)
+  module D =
+  struct 
+    include Lattice.Prod (DF) (DF)
+
+    let join (df1, dnf1) (df2, dnf2) =
+      let df_both = DF.meet df1 df2 in
+      let dnf_any = DF.union dnf1 dnf2 in
+      let df_one = DF.diff (DF.union df1 df2) dnf_any in
+      (DF.union df_both df_one, dnf_any)
+
+    let leq x y = equal (join x y) y
+
+    let widen = join
+    let narrow a b = a
+  end
```

Comment:
I don't think I completely understand this yet... What actually are the components here?

I guess the first one is supposed to be the set of must fresh variables. But what exactly is the second one?

Also, why do we not need to redefine the meet here? Does it just happen to be compatible with the new leq and join?

Reply:
I think the second is the set of variables which may not be fresh, i.e. they have become unfresh by escaping.

`meet` hasn't been redefined because nothing uses it if the analysis itself does not. But it's most likely wrong, so at least should be replaced with
```ocaml
let meet _ _ = assert false
```
Also the `leq` is very naive. I think one could figure out what it should be directly.